File: tests/cases/enforce/no-instantiation/python/inputs/src/orders/checkout.py
from cdec_rules import no_instantiation

from orders.models import AuditEntry, Receipt


@no_instantiation(allow=["AuditEntry"])
class CheckoutService:
    """An orchestrator: it wires collaborators together, it does not build
    them. `AuditEntry` is the one exception, declared in `allow`."""

    def checkout(self, total: float) -> None:
        entry = AuditEntry("checkout")  # allowed by name
        print(entry.action)

    def quick_receipt(self, total: float) -> Receipt:
        """The violation: `Receipt` is not in `allow`."""
        return Receipt(total)

File: tests/cases/enforce/no-instantiation/lua/inputs/src/orders/checkout.lua
local models = require("orders.models")
local Receipt = models.Receipt
local AuditEntry = models.AuditEntry

-- An orchestrator: it wires collaborators together, it does not build them.
---@cdec no_instantiation(allow = {"AuditEntry"})
local CheckoutService = {}
CheckoutService.__index = CheckoutService

-- Allowed by name.
function CheckoutService:checkout()
  return AuditEntry.new("checkout")
end

-- The violation: `Receipt` is not in `allow`.
function CheckoutService:quick_receipt(total)
  return Receipt.new(total)
end

return CheckoutService
